File: AIClient/lib/tools/tool_definition.dart
import 'tool_result.dart';

/// Signature for a tool's execution function.
typedef ToolExecuteFn = Future<ToolResult> Function(Map<String, dynamic> args);

/// Describes a single agentic tool: its name, how the LLM should call it,
/// and the Flutter-side function that executes it.
class ToolDefinition {
  const ToolDefinition({
    required this.name,
    required this.description,
    required this.parametersDoc,
    required this.argsExample,
    required this.execute,
    this.requiresConfirmation = false,
  });

  /// Short snake_case identifier used in <tool_call> JSON.
  final String name;

  /// One-line description shown to the LLM in the system prompt.
  final String description;

  /// Human-readable parameter list for the system prompt (e.g. "on: bool").
  final String parametersDoc;

  /// Example args JSON for the system prompt (e.g. '{"on": true}').
  final String argsExample;

  /// The Flutter function that performs the action and returns a result.
  final ToolExecuteFn execute;

  /// When true, [ChatScreen] shows a confirmation dialog before calling [execute].
  final bool requiresConfirmation;
}

// ── Registry ──────────────────────────────────────────────────────────────────

/// Central catalog of all registered agentic tools.
class ToolRegistry {
  ToolRegistry._();

  static final List<ToolDefinition> _tools = [];

  static void register(ToolDefinition tool) => _tools.add(tool);

  static List<ToolDefinition> get all => List.unmodifiable(_tools);

  static ToolDefinition? find(String name) =>
      _tools.where((t) => t.name == name).firstOrNull;

  static bool get hasTools => _tools.isNotEmpty;

  /// Returns the system-prompt block that lists every registered tool.
  /// Injected before the user's own system prompt on every inference call.
  static String buildSystemPromptBlock() {
    if (_tools.isEmpty) return '';

    final buf = StringBuffer()
      ..writeln('You have access to device hardware tools and a Python execution environment.')
      ..writeln(
          'When you need to use a tool, output EXACTLY this on its own line '
          'and nothing else — wait for the result before continuing:')
      ..writeln('<tool_call>{"name":"TOOL_NAME","args":{...}}</tool_call>')
      ..writeln()
      ..writeln('Available tools:');

    for (final t in _tools) {
      buf
        ..writeln('• ${t.name}(${t.parametersDoc}): ${t.description}')
        ..writeln(
            '  e.g. <tool_call>{"name":"${t.name}","args":${t.argsExample}}</tool_call>');
    }

    return buf.toString().trimRight();
  }
}

File: AIiOS/AIiOS/Inference/ToolDefinition.swift
import Foundation

// MARK: - ToolDefinition

/// Describes a single agentic tool: how the LLM calls it and how the app
/// executes it. Mirrors Flutter's ToolDefinition class.
struct ToolDefinition {
    /// Short snake_case identifier used in <tool_call> JSON.
    let name:            String
    /// One-line description shown to the LLM in the system prompt.
    let description:     String
    /// Human-readable parameter list, e.g. "(no parameters)".
    let parametersDoc:   String
    /// Example args JSON for the system prompt, e.g. "{}".
    let argsExample:     String
    /// The app-side function that performs the action and returns a result string.
    let execute:         ([String: Any]) async -> String
}

// MARK: - ToolRegistry

/// Central catalog of all registered agentic tools.
/// Provides the system-prompt block that tells the LLM what tools it has.
final class ToolRegistry {

    static let shared = ToolRegistry()
    private init() {}

    private var tools: [ToolDefinition] = []

    // ── Registration ──────────────────────────────────────────────────────────

    /// Registers a tool. If a tool with the same name already exists it is
    /// replaced, so calling this from an init method is idempotent.
    func register(_ tool: ToolDefinition) {
        if let idx = tools.firstIndex(where: { $0.name == tool.name }) {
            tools[idx] = tool
        } else {
            tools.append(tool)
        }
    }

    var all:        [ToolDefinition] { tools }
    var hasTools:   Bool             { !tools.isEmpty }

    func find(_ name: String) -> ToolDefinition? {
        tools.first { $0.name == name }
    }

    // ── System-prompt block ───────────────────────────────────────────────────

    /// Returns the block injected at the start of every system prompt so the
    /// LLM knows which tools are available and how to call them.
    func buildSystemPromptBlock() -> String {
        guard !tools.isEmpty else { return "" }

        var buf = """
You have access to device hardware tools.
When you need to use a tool, output EXACTLY one line in this format and nothing else — then wait for the result before continuing:
<tool_call>{"name":"TOOL_NAME","args":{...}}</tool_call>
IMPORTANT: The entire tool_call must be valid JSON on a single line. \
Do NOT use actual newlines inside the tool_call tags.

Available tools:
"""
        for t in tools {
            buf += "\n• \(t.name)(\(t.parametersDoc)): \(t.description)"
            buf += "\n  e.g. <tool_call>{\"name\":\"\(t.name)\",\"args\":\(t.argsExample)}</tool_call>"
        }
        return buf
    }
}
